quinn-proto/src/tests.rs

```diff
@@ -288,47 +330,57 @@ impl TestEndpoint {
-            let (_, ecn, packet) = self.inbound.pop_front().unwrap();
-            if let Some(ch) = self
-                .endpoint
-                .handle(now, remote, ecn, Vec::from(packet).into())
-            {
-                self.accepted = Some(ch);
+
+            for (_, mut events) in self.conn_events.drain() {
+                for event in events.drain(..) {
+                    conn.handle_event(event);
+                }
+            }
+
+            while let Some(event) = conn.poll_endpoint_events() {
+                endpoint_events.push((*ch, event));
+            }
+
+            while let Some(x) = conn.poll_transmit(now) {
+                self.outbound.push_back(x);
+            }
+
+            while let Some(x) = conn.poll_timers() {
+                self.timers[x.timer as usize] = match x.update {
+                    TimerSetting::Stop => {
+                        trace!(
+                            log,
+                            "{side:?} {timer:?} stop",
+                            side = self.side,
+                            timer = x.timer
+                        );
+                        None
+                    }
+                    TimerSetting::Start(time) => {
+                        trace!(
+                            log,
+                            "{side:?} {timer:?} set to expire at {:?}",
+                            time,
+                            side = self.side,
+                            timer = x.timer,
+                        );
+                        Some(time)
+                    }
+                };
             }
         }
-        self.endpoint.handle_endpoint_events();
-        while let Some(x) = self.endpoint.poll_transmit(now) {
-            self.outbound.push_back(x);
-        }
-        while let Some((ch, x)) = self.endpoint.poll_timers() {
-            self.conn = Some(ch);
-            self.timers[x.timer as usize] = match x.update {
-                TimerSetting::Stop => {
-                    trace!(
-                        log,
-                        "{side:?} {timer:?} stop",
-                        side = self.side,
-                        timer = x.timer
-                    );
-                    None
-                }
-                TimerSetting::Start(time) => {
-                    trace!(
-                        log,
-                        "{side:?} {timer:?} set to expire at {:?}",
-                        time,
-                        side = self.side,
-                        timer = x.timer,
-                    );
-                    Some(time)
-                }
-            };
+
+        for (ch, event) in endpoint_events {
+            if let Some(event) = self.handle_event(ch, event) {
+                self.conn_events
+                    .entry(ch)
+                    .or_insert_with(|| VecDeque::new())
+                    .push_back(event);
+            }
         }
```

Comment:
We don't do anything like that for sends and receives, right? We do just one round of processing for each, and then leave the next round for the next call. Why would the internal event handling be substantially different from that model?

Reply:
`EndpointEvent::NeedIdentifiers` can be emitted in response to a receive and will elicit an immediate `NewIdentifiers`, which turns into an immediate send. In the old tokio code, we take care to process it after receive and before send, and in the new tokio code the `mpsc` streams ensures we'll be woken up as needed. Here there's no wakeups, so we must be careful to ensure there's no work left to do before returning from `drive`. Rather than looping as I originally suggested, it's probably sufficient (for now) to reorder things to recv -> process events -> send, since sending never has side effects other than the scheduling of timers.

Reply:
On further consideration, this has to be:
1. recv
2. process connection events (such as `Datagram`)
3. process endpoint events (such as `NeedIdentifiers`)
4. process connection events (such as `NewIdentifiers`)
5. send

(or a loop) to ensure that we don't end up with a client failing to send new identifiers until there's other activity.

Reply:
So I've changed the code to push any `ConnectionEvent`s resulting from `Endpoint::handle_event()` directly into the concerned `Connection`, does that assuage your concerns on this point?

The problem with exactly implementing your proposed change is that the overlapping ownership of the `Endpoint` (for `Endpoint::handle_event()`) and the `Connection`s that are part of it, so the code ends up a bit ugly.

Do you have any concrete further worries here before we can get this PR merged?

Reply:
We'll need to drain `poll_transmit` again to finish, which will set new timers (though in practice we should have just transmitted anyway, so they'll be getting set to the same value again). If the tests are passing I guess we can defer, but a divergence (i.e. letting the new connection IDs packet get stuck in the outbound queue until the something calls `drive` again for an unrelated reason) from realistic behavior is a landmine.

Maybe a more convenient way to structure this would be to iterate over the connections twice? Once to handle incoming packets and timeouts and generate endpoint events, then again after the endpoint events are processed to process any resulting connection events, send packets, and set new timers.